File: Haskell/src/Interface/Arte.hs
module Interface.Arte where

import System.Console.ANSI (clearScreen)
import System.FilePath ((</>))
import System.Directory (doesFileExist)

data ArtType = Derrota | Menu | Jogo | Transicao | Vitoria | Adeus
    deriving (Show, Eq)

-- Nomes dos arquivos de arte
artFileName :: ArtType -> FilePath
artFileName art = case art of
    Derrota    -> "derrota_art.txt"
    Menu       -> "menu_art.txt"
    Jogo       -> "jogo_art.txt"
    Vitoria    -> "vitoria_art.txt"
    Adeus      -> "adeus_art.txt"
    Transicao  -> "" -- Transição pode não ter arte

-- Carrega a arte usando caminho relativo fixo
loadArt :: ArtType -> IO String
loadArt artType = do
    
    let path = "src" </> "Interface" </> "Arte_ascii" </> artFileName artType

    exists <- doesFileExist path
    if exists
        then readFile path
        else do
            putStrLn $ "Aviso: Arquivo de arte não encontrado em '" ++ path ++ "'. Usando arte padrão."
            return $ defaultArt artType

-- Arte padrão caso o arquivo não exista
defaultArt :: ArtType -> String
defaultArt artType = case artType of
    Derrota   -> "DERROTA!"
    Menu      -> "BATALHA NAVAL"
    Jogo      -> "BATALHA NAVAL"
    Vitoria   -> "VITORIA!"
    Adeus     -> "OBRIGADO POR JOGAR!"
    Transicao -> "Carregando..."

-- Exibe arte com limpeza de tela
displayArt :: ArtType -> IO ()
displayArt artType = do
    clearScreen
    if artType == Transicao
        then putStrLn (defaultArt Transicao)
        else loadArt artType >>= putStrLn

File: Prolog/src/interface/arte.pl
:- module(arte, [
    display_art/1,
    clear_screen/0
]).

:- use_module(library(readutil)).
:- use_module(library(filesex)).


art_directory('src/interface/arte_ascii').


art_file(menu,       'menu_art.txt').
art_file(preparacao, 'preparacao_art.txt').
art_file(jogo,       'jogo_art.txt').
art_file(vitoria,    'vitoria_art.txt').
art_file(derrota,    'derrota_art.txt').
art_file(adeus,      'adeus_art.txt').


display_art(ArtType) :-
    load_art(ArtType, Conteudo),
    clear_screen,
    format("~s~n", [Conteudo]).


load_art(ArtType, Conteudo) :-
    art_directory(Dir),
    art_file(ArtType, FileName),
    atomic_list_concat([Dir, '/', FileName], Path),
    (   exists_file(Path)
    ->  read_file_to_string(Path, Conteudo, [])
    ;   default_art(ArtType, Conteudo)
    ).

clear_screen :-
    (   current_prolog_flag(windows, true)
    ->  shell('cls')
    ;   shell('clear')
    ).


default_art(menu,       '=== MENU BATALHA NAVAL ===').
default_art(preparacao, 'Preparando tabuleiros...').
default_art(jogo,       'Batalha Naval em andamento!').
default_art(vitoria,    'PARABÉNS! Você venceu!').
default_art(derrota,    'Você perdeu. Tente novamente!').
default_art(adeus,      'Obrigado por jogar!').
